File: tools.js
// tools.js — ODIN Filesystem Core
// Actions: parse_workspace | read_file | write_file | search_files
const fs   = require('fs');
const path = require('path');

const action     = process.argv[2];
const targetPath = process.argv[3];
const extraArg   = process.argv[4]; // content for write, query for search

const SKIP_DIRS = new Set(['node_modules', '.git', 'venv', '__pycache__', '.cache', 'dist', 'build']);

function out(success, data, error = null) {
  console.log(JSON.stringify({ status: success ? 'SUCCESS' : 'ERROR', payload: data, error }));
  process.exit(success ? 0 : 1);
}

function resolvePath(p) {
  return path.resolve(p.replace(/^~/, process.env.HOME || '/root'));
}

// ── Recursive directory walker ──────────────────────────────────────────────
function walk(dir, depth = 0, maxDepth = 6) {
  if (depth > maxDepth) return [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const results = [];
  for (const e of entries) {
    if (SKIP_DIRS.has(e.name)) continue;
    const full = path.join(dir, e.name);
    if (e.isDirectory()) {
      results.push({ name: e.name, type: 'dir', path: full, children: walk(full, depth + 1, maxDepth) });
    } else {
      results.push({ name: e.name, type: 'file', path: full, ext: path.extname(e.name) });
    }
  }
  return results;
}

// ── Flat file list for token-cheap searching ─────────────────────────────────
function flatFiles(tree, acc = []) {
  for (const node of tree) {
    if (node.type === 'file') acc.push(node.path);
    else if (node.children) flatFiles(node.children, acc);
  }
  return acc;
}

// ── Grep-style content search (no ripgrep required) ──────────────────────────
function searchFiles(dir, query) {
  const tree  = walk(dir);
  const files = flatFiles(tree);
  const hits  = [];
  const re    = new RegExp(query, 'i');

  for (const f of files) {
    // Only search text-like files
    const ext = path.extname(f).toLowerCase();
    if (['.png','.jpg','.gif','.ico','.ttf','.woff','.woff2','.bin','.zip','.7z'].includes(ext)) continue;
    try {
      const lines = fs.readFileSync(f, 'utf-8').split('\n');
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          hits.push({ file: f, line: i + 1, text: lines[i].trim() });
          if (hits.length >= 50) return hits; // cap results
        }
      }
    } catch (_) { /* skip unreadable */ }
  }
  return hits;
}

// ── Dispatch ─────────────────────────────────────────────────────────────────
try {
  if (!targetPath) throw new Error('Missing path argument.');
  const abs = resolvePath(targetPath);

  if (action === 'parse_workspace') {
    if (!fs.existsSync(abs)) throw new Error(`Path not found: ${abs}`);
    const tree  = walk(abs);
    const files = flatFiles(tree);
    out(true, { root: abs, file_count: files.length, tree, flat_paths: files });

  } else if (action === 'read_file') {
    if (!fs.existsSync(abs)) throw new Error(`File not found: ${abs}`);
    out(true, { path: abs, content: fs.readFileSync(abs, 'utf-8') });

  } else if (action === 'write_file') {
    if (!extraArg) throw new Error('Missing content argument.');
    fs.mkdirSync(path.dirname(abs), { recursive: true });
    fs.writeFileSync(abs, extraArg, 'utf-8');
    out(true, `File written: ${abs}`);

  } else if (action === 'search_files') {
    if (!extraArg) throw new Error('Missing search query.');
    if (!fs.existsSync(abs)) throw new Error(`Directory not found: ${abs}`);
    const hits = searchFiles(abs, extraArg);
    out(true, { query: extraArg, hit_count: hits.length, hits });

  } else {
    throw new Error(`Unknown action: ${action}`);
  }

} catch (err) {
  out(false, null, err.message);
}

File: tools.py
# app/tools.py — ODIN Tool Wrappers
# Calls tools.js via subprocess; returns clean strings to the agent loop.

import subprocess
import json
import os
from langchain_core.tools import tool

# tools.js lives in the same directory
_TOOLS_JS = os.path.join(os.path.dirname(__file__), 'tools.js')


def _run_node(action: str, path: str, extra: str = None) -> dict:
    """Execute tools.js and return parsed JSON dict."""
    cmd = ['node', _TOOLS_JS, action, path]
    if extra is not None:
        cmd.append(extra)
    r = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
    try:
        return json.loads(r.stdout)
    except Exception:
        return {'status': 'ERROR', 'payload': None, 'error': r.stdout or r.stderr}


# ── Tool 1: Workspace Parser ─────────────────────────────────────────────────
@tool
def parse_workspace(workspace_path: str) -> str:
    """
    Recursively maps a project directory. Returns a flat list of all file paths
    plus a nested tree. Run this ONCE before any read/write so you know exactly
    what exists and where. Never guess paths — always parse first.
    """
    d = _run_node('parse_workspace', workspace_path)
    if d.get('status') == 'SUCCESS':
        p = d['payload']
        # Return compact summary + flat path list (cheap on tokens)
        flat = '\n'.join(p.get('flat_paths', []))
        return f"ROOT: {p['root']}\nFILES ({p['file_count']}):\n{flat}"
    return f"ERROR: {d.get('error')}"


# ── Tool 2: Search Files ──────────────────────────────────────────────────────
@tool
def search_files(workspace_path: str, query: str) -> str:
    """
    Grep-style search across all text files in workspace_path.
    Use this instead of reading every file — returns only matching lines with
    their file path and line number. Much cheaper than bulk file reads.
    Example: search_files('/home/kwatk/Odin1', 'def run_agent')
    """
    d = _run_node('search_files', workspace_path, query)
    if d.get('status') == 'SUCCESS':
        p = d['payload']
        if not p['hits']:
            return f"No matches for '{query}' in {workspace_path}"
        lines = [f"{h['file']}:{h['line']}  {h['text']}" for h in p['hits']]
        return f"HITS ({p['hit_count']}) for '{query}':\n" + '\n'.join(lines)
    return f"ERROR: {d.get('error')}"


# ── Tool 3: Read File ─────────────────────────────────────────────────────────
@tool
def read_file(file_path: str) -> str:
    """
    Read the full content of a single file. Only use after parse_workspace
    confirms the path exists. Do not use to scan directories.
    """
    d = _run_node('read_file', file_path)
    if d.get('status') == 'SUCCESS':
        return d['payload']['content']
    return f"ERROR: {d.get('error')}"


# ── Tool 4: Write File ────────────────────────────────────────────────────────
@tool
def write_file(file_path: str, content: str) -> str:
    """
    Write or overwrite a file at file_path with content.
    Creates parent directories automatically. Returns SUCCESS or ERROR from OS.
    Never claim a file was written without calling this tool — the OS result
    is the only ground truth.
    """
    d = _run_node('write_file', file_path, content)
    if d.get('status') == 'SUCCESS':
        return f"SUCCESS: {d['payload']}"
    return f"ERROR: {d.get('error')}"
